lib/phoenix_component/macro_component.ex

```diff
@@ -0,0 +1,327 @@
+  @type attributes :: [attribute()]
+  @type children :: [heex_ast()]
+  @type tag_meta :: %{closing: :self | :void}
+  @type heex_ast :: {tag(), attributes(), children(), tag_meta()} | binary()
+  @type transform_meta :: %{env: Macro.Env.t()}
+
+  @callback transform(heex_ast :: heex_ast(), meta :: transform_meta()) ::
+              {:ok, heex_ast()} | {:ok, heex_ast(), data :: term()}
+
+  @doc """
+  Returns the stored data from macro components that returned `{:ok, ast, data}`.
+
+  As one macro component can be used multiple times in one module, the result is a list of all data values.
+
+  If the component module does not have any macro components defined, an empty list is returned.
+  """
+  @spec get_data(module(), module()) :: [term()] | nil
+  def get_data(component_module, macro_component) do
+    if Code.ensure_loaded?(component_module) and
+         function_exported?(component_module, :__phoenix_macro_components__, 0) do
+      component_module.__phoenix_macro_components__()
+      |> Map.get(macro_component, [])
+    else
+      []
+    end
+  end
+
+  @doc false
+  def build_ast([{:tag, name, attrs, tag_meta} | rest], env) do
+    if closing = tag_meta[:closing] do
+      # we assert here because we don't expect any other values
+      true = closing in [:self, :void]
+    end
+
+    meta = Map.take(tag_meta, [:closing])
+    build_ast(rest, [], [{name, token_attrs_to_ast(attrs, env), meta}], env)
+  end
+
+  # recursive case: build_ast(tokens, acc, stack)
+
+  # closing for final stack element -> done!
+  defp build_ast([{:close, :tag, _, _} | rest], acc, [{tag_name, attrs, meta}], _env) do
+    {:ok, {tag_name, attrs, Enum.reverse(acc), meta}, rest}
+  end
+
+  # tag open (self closing or void)
+  defp build_ast([{:tag, name, attrs, %{closing: type}} | rest], acc, stack, env)
+       when type in [:self, :void] do
+    acc = [{name, token_attrs_to_ast(attrs, env), [], %{closing: type}} | acc]
+    build_ast(rest, acc, stack, env)
+  end
+
+  # tag open
+  defp build_ast([{:tag, name, attrs, _tag_meta} | rest], acc, stack, env) do
+    build_ast(rest, [], [{name, token_attrs_to_ast(attrs, env), %{}, acc} | stack], env)
+  end
+
+  # tag close
+  defp build_ast(
+         [{:close, :tag, name, _tag_meta} | tokens],
+         acc,
+         [{name, attrs, meta, prev_acc} | stack],
+         env
+       ) do
+    build_ast(tokens, [{name, attrs, Enum.reverse(acc), meta} | prev_acc], stack, env)
+  end
+
+  # text
+  defp build_ast([{:text, text, _meta} | rest], acc, stack, env) do
+    build_ast(rest, [text | acc], stack, env)
+  end
+
+  # unsupported token
+  defp build_ast([{type, _name, _attrs, meta} | _tokens], _acc, _stack, _env)
+       when type in [:local_component, :remote_component] do
+    {:error, "function components cannot be nested inside a macro component", meta}
+  end
+
+  defp build_ast([{:expr, _, _} | _], _acc, _stack, _env) do
+    raise ArgumentError, "EEx is not currently supported in macro components"
```

Comment:
```suggestion
    # we raise here because we don't have a meta map (line + column) available
    raise ArgumentError, "EEx is not currently supported in macro components"
```